In [ ]:
import tensorflow
from tensorflow import keras
from keras.layers import Dense, Flatten, Conv2D,AveragePooling2D
from keras.models import Sequential
from keras.datasets import mnist
import numpy as np
import pandas as pd

In [ ]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [ ]:
from keras.preprocessing.image import img_to_array, array_to_img
from tensorflow.keras.utils import to_categorical

In [ ]:
X_train_resized = np.zeros((X_train.shape[0], 32, 32, 1))
X_test_resized = np.zeros((X_test.shape[0], 32, 32, 1))

for i in range(X_train.shape[0]):
    X_train_resized[i] = img_to_array(array_to_img(X_train[i].reshape(28, 28, 1)).resize((32, 32)))

for i in range(X_test.shape[0]):
    X_test_resized[i] = img_to_array(array_to_img(X_test[i].reshape(28, 28, 1)).resize((32, 32)))


In [ ]:
X_train = X_train_resized.astype('float32') / 255
X_test = X_test_resized.astype('float32') / 255
Y_train = to_categorical(y_train, 10)
Y_test = to_categorical(y_test, 10)

In [ ]:
X_train.shape

(60000, 32, 32, 1)

In [ ]:
model = Sequential()
model.add(Conv2D(6, kernel_size=(5, 5), activation='tanh', input_shape=(32, 32, 1)))
model.add(AveragePooling2D(pool_size=(2, 2),strides=2))
model.add(Conv2D(16, kernel_size=(5, 5), activation='tanh'))
model.add(AveragePooling2D(pool_size=(2, 2),strides=2))
model.add(Flatten())
model.add(Dense(120, activation='tanh'))
model.add(Dense(84, activation='tanh'))
model.add(Dense(10, activation='softmax'))


In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 28, 28, 6)         156       
                                                                 
 average_pooling2d_2 (Avera  (None, 14, 14, 6)         0         
 gePooling2D)                                                    
                                                                 
 conv2d_3 (Conv2D)           (None, 10, 10, 16)        2416      
                                                                 
 average_pooling2d_3 (Avera  (None, 5, 5, 16)          0         
 gePooling2D)                                                    
                                                                 
 flatten_1 (Flatten)         (None, 400)               0         
                                                                 
 dense_3 (Dense)             (None, 120)              

In [ ]:
model.compile(loss=keras.losses.sparse_categorical_crossentropy, optimizer=keras.optimizers.Adam(), metrics=['accuracy'])

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))


Epoch 1/10
1875/1875 [==============================] - 45s 23ms/step - loss: 0.2236 - accuracy: 0.9335 - val_loss: 0.0990 - val_accuracy: 0.9704
Epoch 2/10
1875/1875 [==============================] - 46s 24ms/step - loss: 0.0832 - accuracy: 0.9736 - val_loss: 0.0625 - val_accuracy: 0.9800
Epoch 3/10
1875/1875 [==============================] - 40s 22ms/step - loss: 0.0539 - accuracy: 0.9829 - val_loss: 0.0590 - val_accuracy: 0.9810
Epoch 4/10
1875/1875 [==============================] - 43s 23ms/step - loss: 0.0418 - accuracy: 0.9869 - val_loss: 0.0487 - val_accuracy: 0.9840
Epoch 5/10
1875/1875 [==============================] - 42s 22ms/step - loss: 0.0342 - accuracy: 0.9891 - val_loss: 0.0478 - val_accuracy: 0.9842
Epoch 6/10
1875/1875 [==============================] - 43s 23ms/step - loss: 0.0275 - accuracy: 0.9910 - val_loss: 0.0576 - val_accuracy: 0.9828
Epoch 7/10
1875/1875 [==============================] - 42s 23ms/step - loss: 0.0237 - accuracy: 0.9922 - val_loss: 0.0472 -

In [ ]:
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print("Test Loss:", test_loss)
print("Test Accuracy:", test_accuracy)

313/313 [==============================] - 3s 8ms/step - loss: 0.0541 - accuracy: 0.9852
Test Loss: 0.05405160039663315
Test Accuracy: 0.9851999878883362


In [ ]:
model.save("my_LetNet5.h5")

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [ ]:
y_pred = model.predict(X_test)

313/313 [==============================] - 3s 9ms/step


In [ ]:
from sklearn.metrics import classification_report
y_pred_classes = np.argmax(y_pred, axis=1)
report = classification_report(y_test, y_pred_classes)
print(report)

              precision    recall  f1-score   support

           0       0.99      0.99      0.99       980
           1       0.99      1.00      0.99      1135
           2       0.98      0.99      0.98      1032
           3       0.98      0.99      0.98      1010
           4       0.98      0.98      0.98       982
           5       0.98      0.99      0.99       892
           6       0.99      0.98      0.99       958
           7       0.99      0.98      0.98      1028
           8       0.99      0.97      0.98       974
           9       0.98      0.98      0.98      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



Accuracy: 98.52% Test Loss:0.054 Training Time: Faster than CNN's Hyperparameters: C (Regularization parameter): Controls the trade-off between achieving a low error on training data and minimizing the complexity of the model. 10 Gamma (Kernel coefficient): Defines how far the influence of a single training example reaches, impacting the decision boundary shape in the feature space. 0.001 Special features : handling non-linear classification tasks without explicitly transforming the input data. generally robust to noise in the data and outliers